implementation/64103 Extend API V3 to cover Emoji reactions on work package comments (read) and extend documentation

akabiru opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

https://community.openproject.org/work_packages/64103

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

- [x] Add _embedded emoji reactions to Activities
- [x] Add `GET /api/v3/activities/:id/emoji_reactions`
- [x] Add `GET /api/v3/work_packages/:id/activities_emoji_reactions` *collection endpoint for all work package comments emoji reaction
- [x] Add API documentation

_Visit /api/docs_

<img width="1461" alt="Screenshot 2025-05-21 at 5 40 36 PM" src="https://github.com/user-attachments/assets/3bc1bfee-4b12-4898-bde3-0afda2059f49" />


# Merge checklist

- [x] Added/updated tests
- [x] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

akabiru committed:
implementation/64103 Extend API V3 to cover Emoji reactions on work package comments (read) and extend documentation

https://community.openproject.org/work_packages/64103

akabiru committed:
Use the correct path for work package comments emoji reactions

akabiru committed:
Add convenience method for retrieving emojis

akabiru committed:
Add /api/v3/activities/:id/emoji_reactions requests specs

akabiru committed:
Add /api/v3/work_packages/:id/activities_emoji_reactions request specs

akabiru committed:
Add /api/v3/activities/:id emoji reaction embed request spec

akabiru committed:
Add emoji reaction representer unit spec

akabiru committed:
Reduce no. of assertions

Rely on the representer for full representation unit tests

akabiru committed:
Avoid N+1 for emoji reaction reactable reads

akabiru committed:
Add emoji reactions openapi specification

brunopagno commented on lib/api/v3/emoji_reactions/emoji_reaction_representer.rb:
```diff
@@ -0,0 +1,116 @@
+        def initialize(model, current_user:, embed_links: false)
+          raise "no represented object passed" if model_required? && model.nil?
+
+          # EmojiReaction#updated_at is set to the first created_at timestamp of the reaction.
+          #
+          # This is done to ensure that the updated_at timestamp is always present for caching purposes.
+          model.updated_at = model.first_created_at
```
If this is necessary, shouldn't this be done during persisting instead of when rendering the result?

I'm not sure what are the implications of such change, though, so all good if this is the way.

akabiru replied:
Yes, this is an odd one- the model at this stage is actually _not_ a "raw" "emoji_reaction" rather a grouped emoji reaction (EmojiReaction<reaction,reactionsCount,reactingUsers,firstCreatedAt>) so there's really no updated_at timestamp, which is needed for caching.

We therefore only set it purely for read purposes, as this is a representer- persisting will never be called. And even if "save" is called for some odd reason, it would result in a no-op.

Context: https://github.com/opf/openproject/blob/dev/spec/models/concerns/reactable_spec.rb#L135

akabiru replied:
I had previously tried overriding the CachedRepresenter to make use of the "first_created_at" value as it's key, maybe that's the more saner and clearer approach. I'll look into it :)

akabiru replied:
tl;dr - we should actually remove the cache support from the emoji reactions representer.

After digging into OpenProject::Cache::CacheKey which extends ActiveSupport::Cache, the cache key resolution is p[retty much coupled to an updated_at column presence](https://github.com/rails/rails/blob/ad858b91a9a4bc94950708955e44c654a1f3789b/activerecord/lib/active_record/integration.rb#L97).

However, since this is a grouped collection using the firstReactedAt as the cacheKey would result in stale results, as that is the timestamp of the first reaction, and any more added would still maintain the same key- until the first reaction is changed. Which is not what we want. I'm in favour of removing the cache alltogether.

brunopagno reviewed (approved):
Loooks good to me. I haven't checked locally, though, but the code looks sane.

Left a comment if you want to dig stuff.

akabiru committed:
Remove representer caching from emoji reactions representer

Given the EmojiReactionsRepresenter do not present an #updated_at column, it would lead to stale
caches if we used the firstReactionAt as the cache key.
